File: pset6/credit/credit.py
from cs50 import get_int


def main():

    # Get number from user
    number = get_positive_int("Number: ")

    # Initialise variables for convenient use later
    digit_count = count_digits(number)
    digit12 = first_2_digits(number)

    # Check if luhn algorithm test result is true, else print invalid
    if luhn_algorithm(number):

        # Checking various characteristics of american express, mastercard and visa
        if digit_count == 15 and (digit12 == 34 or digit12 == 37):
            print("AMEX")
            return
        if digit_count == 16 and digit12 >= 51 and digit12 <= 55:
            print("MASTERCARD")
            return
        if (digit_count == 13 or digit_count == 16) and int(digit12 / 10) == 4:
            print("VISA")
        else:
            print("INVALID")
    else:
        print("INVALID")


# Luhn's Algorithm
def luhn_algorithm(number):

    # Initialise to zero
    non_multiplied_sum = 0
    multiplied_sum = 0

    while True:
        # Summing all odd-placed digits
        non_multiplied_sum += int(number) % 10
        number /= 10
        if number < 1:
            break

        # Summing all even-placed or every other digit, but first checking if multiplied digit is single or double digit
        multiplied_digit = (int(number) % 10) * 2
        number /= 10
        if multiplied_digit < 10:
            multiplied_sum += multiplied_digit
        else:
            multiplied_sum += multiplied_digit % 10
            multiplied_sum += 1
        if number < 1:
            break

    # Checking if last digit of sum is 0 and return true or false
    return (multiplied_sum + non_multiplied_sum) % 10 == 0


# Return number of digits in number
def count_digits(number):
    i = 0
    while number >= 1:
        number /= 10
        i += 1
    return i


# Return first 2 digits in number
def first_2_digits(number):
    n = int(number / 10**(count_digits(number) - 2))
    return n


# Ensure input is positive integer
def get_positive_int(prompt):
    while True:
        n = get_int(prompt)
        if n > 0:
            return n
            break


if __name__ == "__main__":
    main()

File: pset1/credit/credit.c
#include <cs50.h>
#include <stdio.h>

long GetPositiveInt(void);
int DigitCount(long j);
bool LuhnAlgorithm(long number);
int Digit1(long number);
int Digit12(long number);

int main(void)
{
    //Prompt to get positive integer
    long number = GetPositiveInt();

    //Count number of digits
    int digit_count = DigitCount(number);

    //Validating against LuhnAlgorithm
    bool valid = LuhnAlgorithm(number);

    //Obtaining first digit
    int digit1 = Digit1(number);

    //obtaining first 2 digits
    int digit12 = Digit12(number);

    //Categorising invalids and various cards
    if (valid == true)
    {
        if (digit_count == 15 && (digit12 == 34 || digit12 == 37))
        {
            printf("AMEX\n");
        }
        else if (digit_count == 16 && (digit12 >= 51 && digit12 <= 55))
        {
            printf("MASTERCARD\n");
        }
        else if (digit1 == 4 && (digit_count == 13 || digit_count == 16))
        {
            printf("VISA\n");
        }
        else
        {
            printf("INVALID\n");
        }
    }
    else
    {
        printf("INVALID\n");
    }

}

//Abstraction: Prompt to get positive integers
long GetPositiveInt(void)
{
    long n;
    do
    {
        n = get_long("Number: ");
    }
    while (n < 1);
    return n;
}

//Abstraction: Count number of digits
int DigitCount(long j)
{
    int n = 0;
    for (long i = j; i >= 1; i /= 10)
    {
        n++;
    }
    return n;
}

//Abstraction: Validation Using Luhn's Algorithm
bool LuhnAlgorithm(long number)
{
    int non_multiplied_sum = 0;
    int multiplied_sum = 0;
    int n = 0;
    for (long i = number; i >= 1; i /= 10)
    {
        int digit = i % 10;
        //Case when digit is not to be mulitplied
        if (n % 2 == 0)
        {
            non_multiplied_sum += digit;
        }
        //Case when digit is to be multiplied by 2
        else
        {
            int multiplied2 = digit * 2;
            //Cases when digit to be multiplied by 2 may be double or single digit or 0
            switch (DigitCount(multiplied2))
            {
                case 2:
                    multiplied_sum += multiplied2 % 10 + (multiplied2 / 10) % 10;
                    break;
                case 1:
                case 0:
                    multiplied_sum += multiplied2;
            }
        }
        n++;
    }
    //To Check if last digit of the sum is indeed 0
    if ((non_multiplied_sum + multiplied_sum) % 10 == 0)
    {
        return true;
    }
    else
    {
        return false;
    }
}

//Abstraction: Obtaining first digit
int Digit1(long number)
{
    int n;
    for (long i = number; i >= 1; i /= 10)
    {
        n = i % 10;
    }
    return n;
}

//Abstraction: Obtaining first 2 digits
int Digit12(long number)
{
    int n;
    for (long i = number; i >= 10; i /= 10)
    {
        n = i % 100;
    }
    return n;
}